src/SAWScript/Proof.hs

```diff
@@ -1141,56 +1825,109 @@ tacticIntro :: (F.MonadFail m, MonadIO m) =>
+            (Just h', Just g') ->
+              do g'' <- liftIO (Prop <$> (scEqTrue sc =<< scImplies sc h' g'))
+                 let sqt' = GoalFocusedSequent hs (FB gs1 g'' gs2)
+                 let goal' = goal{ goalSequent = sqt' }
+                 return ((), mempty, [goal'], updateEvidence (NormalizeSequentEvidence sqt'))
+
+            _ -> fail "goal_revert_hyp: expected EqTrue props"
+        _ -> fail "goal_revert_hyp: not enough hypotheses"
+    _ -> fail "goal_revert_hyp: conclusion focus required"
+
+
+-- | Attempt to prove a goal by applying a local hypothesis.  Any hypotheses of
+--   the applied proposition will generate additional subgoals.
+tacticApplyHyp :: (F.MonadFail m, MonadIO m) => SharedContext -> Integer -> Tactic m ()
+tacticApplyHyp sc n = Tactic \goal ->
+  case goalSequent goal of
+    UnfocusedSequent{} -> fail "apply hyp tactic: focus required"
+    HypFocusedSequent{} -> fail "apply hyp tactic: cannot apply in a hypothesis"
+    GoalFocusedSequent hs (FB gs1 g gs2) ->
+      case genericDrop n hs of
+        (h:_) ->
+          liftIO (goalApply sc h g) >>= \case
+            Nothing -> fail "apply hyp tactic: no match"
+            Just newterms ->
+              let newgoals =
+                    [ goal{ goalSequent = GoalFocusedSequent hs (FB gs1 p gs2)
+                          , goalType = goalType goal ++ ".subgoal" ++ show i
+                          }
+                    | Right p <- newterms
+                    | i <- [0::Integer ]
+                    ] in
+              return ((), mempty, newgoals, \es -> ApplyHypEvidence n <$> processEvidence newterms es)
+        _ -> fail "apply hyp tactic: not enough hypotheses"
+
+ where
+   processEvidence :: [Either Term Prop] -> [Evidence] -> IO [Either Term Evidence]
+   processEvidence (Left tm : xs) es     = (Left tm :) <$> processEvidence xs es
+   processEvidence (Right _ : xs) (e:es) = (Right e :) <$> processEvidence xs es
+   processEvidence []             []     = pure []
+   processEvidence _ _ = fail "apply hyp tactic failed: evidence mismatch"
+
 
 -- | Attempt to prove a goal by applying the given theorem.  Any hypotheses of
 --   the theorem will generate additional subgoals.
 tacticApply :: (F.MonadFail m, MonadIO m) => SharedContext -> Theorem -> Tactic m ()
 tacticApply sc thm = Tactic \goal ->
-  liftIO (goalApply sc (thmProp thm) goal) >>= \case
-    Nothing -> fail "apply tactic failed: no match"
-    Just newterms ->
-      let newgoals =
-            [ goal{ goalProp = p, goalType = goalType goal ++ ".subgoal" ++ show i }
-            | Right p <- newterms
-            | i <- [0::Integer ..]
-            ] in
-      return ((), mempty, newgoals, \es -> ApplyEvidence thm <$> processEvidence newterms es)
+  case sequentState (goalSequent goal) of
+    Unfocused -> fail "apply tactic: focus required"
+    HypFocus _ _ -> fail "apply tactic: cannot apply in a hypothesis"
+    GoalFocus gl mkSqt ->
```

Comment:
Maybe? I'm always on the fence about this sort of thing. In this case, I think the direct pattern match may be more clear, actually.

Reply:
Fair enough. If nothing else, it might be worth factoring out the error message so that you don't have to repeatedly type `focus required` and `cannot apply in a hypothesis` everywhere.